public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts

```diff
@@ -169,6 +170,42 @@ export class DashboardScenePageStateManager extends StateManagerBase<DashboardSc
     throw new Error('Snapshot not found');
   }
 
+  public restoreDashboardStateFromLocalStorage(dashboard: DashboardScene) {
+    const preservedUrlState = store.get(getPreservedSceneURLStateKey());
+    if (preservedUrlState) {
+      const preservedUrlStateJSON = preservedUrlState ? JSON.parse(preservedUrlState) : {};
+
+      let preservedQueryParams: UrlQueryMap = {};
+
+      for (const [key, value] of Object.entries(preservedUrlStateJSON)) {
+        // restore non-variable query params
+        if (!key.startsWith('var-')) {
+          preservedQueryParams[key] = value as any;
+          continue;
+        }
+        // restore variable query params if a variable exists in the target dashboard
+        if (dashboard.state.$variables?.getByName(key.replace('var-', ''))) {
+          preservedQueryParams[key] = value as any;
+        }
+      }
+
+      const currentQueryParams = locationService.getLocation().search;
+      let nextQueryParams = currentQueryParams;
+      if (currentQueryParams) {
+        nextQueryParams += '&' + urlUtil.renderUrl('', preservedQueryParams).slice(1);
+      } else {
+        nextQueryParams = urlUtil.renderUrl('', preservedQueryParams);
+      }
```

Comment:
this part feels a bit messy, would it not be easier create a URLSearchParams at the start of this function (one for the current and one for the persisted). 

Then loop through the persisted one, ignore duplicates & validate the variable exists,